src/BinSkim.Rules/PERules/BA2004.EnableSecureSourceCodeHashing.cs

```diff
@@ -134,6 +142,22 @@ public void AnalyzeNativeBinaryAndPdb(BinaryAnalyzerContext context)
                         "native"));
         }
 
+        private void GenerateCompilandsAndLog(BinaryAnalyzerContext context, List<ObjectModuleDetails> compilandsWithOneOrMoreInsecureFileHashes)
+        {
+            string compilands = compilandsWithOneOrMoreInsecureFileHashes.CreateOutputCoalescedByLibrary();
+
+            //'{0}' is a native binary that links one or more object files which were hashed
+            // using an insecure checksum algorithm (MD5). MD5 is subject to collision attacks
+            // and its use can compromise supply chain integrity. Pass '/ZH:SHA-256' on the
+            // cl.exe command-line to enable secure source code hashing. The following modules
+            // are out of policy: {1} 
+            context.Logger.Log(this,
+                RuleUtilities.BuildResult(FailureLevel.Warning, context, null,

```

Comment:
>Warning [](start = 55, length = 7)

Almost done! You need to pass the failure level into this helper, Error for binary compilands, Warning for libray. :) Then see what test breaks occur. Run UpdateBaselines.ps1 in the driver functionals directory and review changes.

Reply:
oops after i create the function, i forgot to add the level. thanks

---
In reply to: [554458619](https://github.com/microsoft/binskim/pull/328#discussion_r554458619) [](ancestors = 554458619)